### ISSUE TITLE
Dataset access speed with [::N] vs [...][::N]

### ISSUE THREAD

**Issue by kutoga** [h5py/h5py#1832]
* Operating System: Ubuntu 18.04
 * Python version: 3.6.9
 * Where Python was acquired: System python
 * h5py version: 3.1.0
 * HDF5 version: 1.12.0

Hi

I've just noticed that the access with `[::N]` is quite consistently slower than with `[...][::N]`.

Usually, loading the complete dataset into memory is not a big problem, except for when it is huge, so `[::N]` would be the much better choice.

An example:
```python
import numpy as np
import h5py

# Create a dataset
a = np.random.random(size=(100000, 10000))
h5f = h5py.File('/home/user/tmp_test.h5', 'w')
h5f.create_dataset('dataset_1', data=a, chunks=(1000, 1000))
h5f.close()

# Read the dataset
h5f = h5py.File('/home/user/tmp_test.h5', 'r', rdcc_nbytes=1024 * 1024 * 1024)

# Read using [::2]
a = None
start = time.time()
a = h5f["dataset_1"][::2]
end = time.time()
print(end - start) # Prints "64.27390313148499"

# Read using [...][::2]
a = None
start = time.time()
a = h5f["dataset_1"][...][::2]
end = time.time()
print(end - start) # Prints "5.501309871673584"

h5f.close()
```

The smaller `N` gets, the bigger is the difference. Except for `N=1` which is as fast as `[...]`. I really would like to access with `[::2]`, but the speed penalty hurts. Maybe I do something wrong? I've played around with the cache, but it did not seem to help too much.

My access pattern is always the same: `[::N]` with N which is a value between 1 and 30. The dataset is only read once. Maybe there is a better way to access the data for this use-case? It also has to be assumed that the complete dataset does not fit into memory. The chunk size can be assumed to be optimized for compression purposes.

Thanks:)

**Comment by takluyver** [comment]
It wouldn't surprise me if it was a bit slower. With chunked data, it always reads entire chunks, and with your example it can't skip any chunks. Then HDF5 has to copy half of that data to another location in memory. Loading the whole dataset and then slicing it in NumPy avoids a memory copy, because NumPy will just give you a view. For a fairer comparison, call `np.ascontiguousarray()` on it after reading.

But I don't think that should make as big a difference as you show. We've run into similar things in the past - it seems that it's easy to stray off HDF5's optimised read pathway and land on some slower fallback.

**Comment by tacaswell** [comment]
https://github.com/h5py/h5py/blob/aab35a17dd777486634536c254708200b7d3ab19/h5py/_hl/dataset.py#L689-L795

is the relevant code on our side.  I have vague memories of performance issues arising both from time spent on the Python side building the hyperslab in a really expensive way and time spent on the libhdf5 side due to hyper slabs expressed in a inefficient ways.

**Comment by kutoga** [comment]
@takluyver 
> It wouldn't surprise me if it was a bit slower. With chunked data, it always reads entire chunks, and with your example it can't skip any chunks. Then HDF5 has to copy half of that data to another location in memory. Loading the whole dataset and then slicing it in NumPy avoids a memory copy, because NumPy will just give you a view. For a fairer comparison, call `np.ascontiguousarray()` on it after reading.
> 
> But I don't think that should make as big a difference as you show. We've run into similar things in the past - it seems that it's easy to stray off HDF5's optimised read pathway and land on some slower fallback.

Good point. I've added a call to `np.ascontiguousarray()`, but as you already guessed, the difference is not significant.

@tacaswell 
> is the relevant code on our side. I have vague memories of performance issues arising both from time spent on the Python side building the hyperslab in a really expensive way and time spent on the libhdf5 side due to hyper slabs expressed in a inefficient ways.

I had a quick look at the code and I'll have one again, but I did not see something obvious. But then it seems anyway not too easy to solve this?

If there is no solution to quickly access every N-th element (in a memory-friendly way), it's also okay for me. I just thought that I maybe use something wrong.

**Comment by takluyver** [comment]
Interestingly, if I use `read_direct()` to read every second row of the dataset into every second row of an array, it's a similar speed to reading the whole thing and then slicing it:

```python
start = time.time()
ds = h5f["dataset_1"]
a = np.zeros(ds.shape, dtype=ds.dtype)
ds.read_direct(a, np.s_[::2], np.s_[::2])
print(a[:10, :10])
end = time.time()
print("read_direct", end - start)
```

Of course, this doesn't really solve anything. But if fits with my notion that there's some optimised fast pathway for reading, which is used when the source and the destination are more similar. I guess the next step would be to see if we can reproduce it in C. The [HDF5 example programs](https://github.com/HDFGroup/hdf5/tree/develop/examples) can be a useful starting point for that kind of thing.
